### DISCUSSION TITLE
three-point-oh-point-oh

### DISCUSSION THREAD

**Discussion by dankamongmen** [dankamongmen/notcurses#2230] [General]
I want to get 3.0.0 out before the end of the year, and with my dayjob's recent gentle insistence that I do more paying work, dev time is at a premium. The importance of the "3.0.0" moniker is the SONAME/ABI version bump; I frankly wouldn't mind releasing 3.0.0 today with all of #1777 resolved.

The one thing I want is for ABI3 to be truly stable, which requires having thought through (if not necessarily implemented all the way) all desirable expansions. The last things I see maybe requiring API changes are #2184, #1682, and #1444:

* #1444: use kitty graphics' z-axis to have text above a graphic; right now, text obliterates graphics
* #1682: smooth bitmap movement, which pretty much requires being able to specify arbitrary pixel locations, either as true absolutes or relative to a cell's origin
* #2184: disambiguate repeat/press/release for input

i think we've made the necessary changes for the last, and the first never really needed changes. the second will probably require an expansion of `struct ncvisual_options`.

is there anything else people think needs resolution for ABI3? any big dreams, perhaps not even yet recorded as bugs, which you'd like to see down the road?

i turn 41 on tuesday, and that was the original target date for 3.0.0. it's not gonna happen, but i definitely want it out in 2021.

**Comment by joseluis** [comment]
what about renaming `ncplane_pixelgeom` to `ncplane_pixel_geom`, for consistency with `ncvisual_blitter_geom` also?

**Reply by joseluis** [reply]
in the same venue renaming `ncplane_moverel` to `ncplane_move_rel` would make it consistent  with the rest of the move methods (`_yx`, `_above`, `_top`, ...).

**Reply by dankamongmen** [reply]
i love it; throw this into #1777 please

**Reply by joseluis** [reply]
great, I've just added them to the list

**Comment by joseluis** [comment]
what about a new function `putstr_aligned_stained`?

**Reply by joseluis** [reply]
BTW there's no mention at all in [notcurses_output.3](https://notcurses.com/notcurses_output.3.html) of what the `_stained` methods do.

**Comment by joseluis** [comment]
happy birthday man! 🎂

**Reply by dankamongmen** [reply]
> happy birthday man!

thanks kindly! i have celebrated it thus far by hopefully solving the last bug in the automata construction. later shall come the slaughter of a large dancing bear. by this evening's end i hope to look like my fellow Samoan in the bathtub. 

https://www.youtube.com/watch?v=1RBwoUbvxx0

**Comment by dankamongmen** [comment]
i'm thinking i might go ahead and start the 3.0.0 branch.

**Comment by joseluis** [comment]
There's another inconsistency that has always bothered me.

The `notcurses_render` function renders and rastarizes the standard plane, but there's no equivalent function for an ncplane, you have to call both `ncplane_render` & `ncplane_rasterize`.

Ideally the same function name should do the same operation in both contexts.

A possible solution could be adding `ncplane_display` for the render+raster operation, and deprecating notcurses_render, renaming it to  `notcurses_display`? Or maybe there's a better function name for that.
